### PR TITLE
Normalize sampled MCDA weights

### LINKED ISSUE #559: Add MCDA weighting normalisation to scenario sampling post-processing
Currently MCDA weighting parameters are sampled from within uniform distributions and then normalised inside `guided_site_selection` so that the sum of weightings for each of the intervention decision matrices is 1. This should instead be part of the post-processing of scenario samples prior to running scenarios in the ecological model, so that the normalisation is not repeated multiple times within `guided_site_selection` when running scenarios and the inputs reflect the magnitude of the parameters as they are used within the model.

### LINKED ISSUE #590: Tests raising error
I run `(ADRIA) pkg> test` and got the following error:

```
Running 32 scenarios for RCP 45   6%|█████████                                                                                                                           |  ETA: 0:08:25

ERROR: LoadError: type LOPCOWResult has no field bestIndex
Stacktrace:
  [1] getproperty
    @ .\Base.jl:37 [inlined]
  [2] retrieve_ranks(S::Matrix{Float64}, site_ids::Vector{Float64}, weights::Vector{Float64}, mcda_func::Type{JMcDM.LOPCOW.LOPCOWMethod})
    @ ADRIA.decision C:\Users\pribeiro\AIMS\Code\ADRIA.jl\src\decision\dMCDA.jl:275
  [3] rank_sites!(S::Matrix{Float64}, weights::Vector{Float64}, rankings::Matrix{Int64}, n_site_int::Int64, mcda_func::Type{JMcDM.LOPCOW.LOPCOWMethod}, rank_col::Int64)
    @ ADRIA.decision C:\Users\pribeiro\AIMS\Code\ADRIA.jl\src\decision\dMCDA.jl:228
  [4] guided_site_selection(d_vars::ADRIA.decision.DMCDA_vars, alg_ind::Int64, log_seed::Bool, log_shade::Bool, pref_seed_locs::Vector{Int64}, pref_shade_locs::Vector{Int64}, rankings_in::Matrix{Int64}, in_conn::Vector{Float64}, out_conn::Vector{Float64}, strong_pred::Vector{Int64}; methods_mcda::Vector{Any})
    @ ADRIA.decision C:\Users\pribeiro\AIMS\Code\ADRIA.jl\src\decision\dMCDA.jl:668
  [5] guided_site_selection
    @ C:\Users\pribeiro\AIMS\Code\ADRIA.jl\src\decision\dMCDA.jl:547 [inlined]
  [6] run_model(domain::ADRIADomain{NamedDims.NamedDimsArray{(:Source, :Receiving), Float64, 2, AxisKeys.KeyedArray{Float64, 2, Matrix{Float64}, Tuple{Vector{String}, Vector{String}}}}, NamedDims.NamedDimsArray{(:species, :sites), Float32, 2, AxisKeys.KeyedArray{Float32, 2, Matrix{Float32}, Tuple{UnitRange{Int64}, Vector{String}}}}, DataFrame, Matrix{Float64}, NamedDims.NamedDimsArray{(:timesteps, :sites, :scenarios), Float32, 3, AxisKeys.KeyedArray{Float32, 3, Array{Float32, 3}, Tuple{UnitRange{Int64}, Vector{String}, UnitRange{Int64}}}}, NamedDims.NamedDimsArray{(:timesteps, :sites, :scenarios), Float32, 3, AxisKeys.KeyedArray{Float32, 3, Array{Float32, 3}, Tuple{UnitRange{Int64}, Vector{String}, UnitRange{Int64}}}}}, param_set::NamedDims.NamedDimsArray{(:factors,), Float64, 1, AxisKeys.KeyedArray{Float64, 1, SubArray{Float64, 1, Matrix{Float64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}, Base.RefValue{Vector{String}}}}, corals::DataFrame, cache::NamedTuple{(:fec_all, :fec_scope, :recruitment, :dhw_step, :cov_tmp, :depth_coeff, :site_area, :site_k_area, :wave_damage, :dhw_tol_mean_log), Tuple{Matrix{Float64}, Matrix{Float64}, Matrix{Float64}, Vector{Float64}, Matrix{Float64}, Vector{Float64}, Matrix{Float64}, Matrix{Float64}, Array{Float64, 3}, Array{Float64, 3}}})
    @ ADRIA C:\Users\pribeiro\AIMS\Code\ADRIA.jl\src\scenario.jl:659

...
```

### PR DESCRIPTION
**Pull request by Rosejoycrocker**
Incorporates changes allowing MCDA weights for fogging and seeding decision matrices to be normalised at the parameter sampling step so that parameter inputs reflect their magnitude as used inside ADRIA.

The heat stress and wave stress weights are also split into separate weights for seeding and fogging so that normalisation can be done correctly (otherwise normalised values for seeding will be overwritten by normalised values for shading and these will not be the same).

Tests added to check normalisation is working as expected.

Note the growth.jl test is failing but this is fixed in PR #573

Closes #559.

Closes #590

### CODE REVIEW

File: src/decision/CriteriaWeights.jl [hunk 1]
```diff
@@ -1,28 +1,51 @@
...
+    shade_heat_stress::P = Param(
+        1.0;
+        ptype="real",
+        bounds=(0.0, 1.0),
+        default_bounds=(0.0, 1.0),
+        dists="unif",
+        criteria_keywords=(:fog, :weight),
+        name="Shade Heat Stress",
+        description="Importance of avoiding heat stress when fogging. Higher values places more weight on areas with low heat stress.",
     )
```
Comment (ConnectedSystems) [inline src/decision/CriteriaWeights.jl:41]: Description and criteria keywords mention `fog`, but factor and field names use "shade" ?

File: src/decision/CriteriaWeights.jl [hunk 2]
```diff
@@ -1,28 +1,51 @@
...
+Base.@kwdef struct CriteriaWeights{P,P2,P3,N} <: EcoModel
+    seed_wave_stress::P = Param(
         1.0;
         ptype="real",
         bounds=(0.0, 1.0),
         default_bounds=(0.0, 1.0),
         dists="unif",
-        name="Wave Stress",
-        description="Importance of avoiding wave stress. Higher values places more weight on areas with low wave stress.",
+        criteria_keywords=(:seed, :weight),
```
Comment (ConnectedSystems) [inline src/decision/CriteriaWeights.jl:8]: Could you elaborate on what the `:weight` keyword is used for? The struct is ostensibly for criteria weights, so are we adding a `:weight` keyword for all parameters here?

File: src/ecosystem/corals/Corals.jl [hunk 3]
```diff
@@ -296,7 +296,10 @@ function create_coral_struct(bounds::Tuple{Float64,Float64}=(0.9, 1.1))::Nothing
         bounds=(0.25, 0.5),
         default_bounds=(0.25, 0.5),
         dists="unif",
-        name="Heritability", description="Heritability of DHW tolerance.")
+        criteria_keywords=(""),
+        name="Heritability",
+        description="Heritability of DHW tolerance.",
+    )
```
Comment (ConnectedSystems) [inline src/ecosystem/corals/Corals.jl:302]: A bit late now but you could have created your own separate constructor that defines a default value for these additional fields (`criteria_keywords`, `default_bounds`, etc) so you don't have to go around adding these fields to all use of `Param` types.

Leave it as is - we'll clean it up later.

File: src/io/sampling.jl [hunk 4]
```diff
@@ -43,6 +46,15 @@ function adjust_samples(spec::DataFrame, df::DataFrame)::DataFrame
     not_fogged = (df.fogging .== 0) .& (df.SRM .== 0)
     df[not_fogged, contains.(names(df), "shade_")] .= 0.0
 
+    # Normalize MCDA weights for fogging scenarios
+    df[.!(not_fogged) .& (df.guided .> 0), weights_fog_crit.fieldname] .= mcda_normalize(
+        df[.!(not_fogged) .& (df.guided .> 0), weights_fog_crit.fieldname],
+    )
```
Comment (ConnectedSystems) [inline src/io/sampling.jl:52]: SRM is applied on a regional basis at the moment - shouldn't `not_fogged` just check if `fogging` == 0?

File: src/decision/CriteriaWeights.jl (hunk 2 above)
Comment (ConnectedSystems) [reply src/decision/CriteriaWeights.jl:8]: I gather it's to distinguish with `threshold` values - okay leave this as is, we'll clean up later as part of a larger restructure.

**Review by ConnectedSystems — CHANGES_REQUESTED** [review]
Just one question from me.

File: src/decision/CriteriaWeights.jl (hunk 1 above)
Comment (Rosejoycrocker) [reply src/decision/CriteriaWeights.jl:41]: Yeah, I originally had "shade" as the descriptor but really it's fogging so I changed it. I can change it back to shade for consistency, or I can change the parameters labeled with "shade" to be labeled with "fog", which will require some other changes throughout the code. What do you think?

File: src/io/sampling.jl (hunk 4 above)
Comment (Rosejoycrocker) [reply src/io/sampling.jl:52]: good point!

File: src/decision/CriteriaWeights.jl (hunk 1 above)
Comment (ConnectedSystems) [reply src/decision/CriteriaWeights.jl:41]: Which is more specific to what is being represented?

Comment (Rosejoycrocker) [reply src/decision/CriteriaWeights.jl:41]: "fog" is more relevant, I can change it to that. Is it ok to also change the order of some parameter names in this PR while I'm at it? Some have seed/shade at the end, some at the start etc.

Comment (ConnectedSystems) [reply src/decision/CriteriaWeights.jl:41]: Yes?

**Comment by Rosejoycrocker** [comment]
I think I've addressed your comments @ConnectedSystems . Tests are all passing on my end. For consistency with changing "shade" to "fog" in `CriteriaWeights.jl` I also changed "shade" to "fog" in other relevant places inside the MCDA, such as the weight names and `pref_shade_sites`. `shade_locs` is also now `fog_locs` in `scenario.jl` and I renamed `create_shade_matrix` to `create_fog_matrix`.

File: src/io/sampling.jl [hunk 5]
```diff
@@ -41,7 +44,16 @@ function adjust_samples(spec::DataFrame, df::DataFrame)::DataFrame
...
+    df[not_fogged, contains.(names(df), "fog_") .& contains.(names(df), "shade_")] .= 0.0
+
+    # Normalize MCDA weights for fogging scenarios
+    df[(df.fogging .> 0.0) .& (df.guided .> 0), weights_fog_crit.fieldname] .= mcda_normalize(
+        df[(df.fogging .> 0.0) .& (df.guided .> 0), weights_fog_crit.fieldname],
+    )
+    # Normalize MCDA weights for seeding scenarios
+    df[.!(not_seeded) .& (df.guided .> 0), weights_seed_crit.fieldname] .= mcda_normalize(
+        df[.!(not_seeded) .& (df.guided .> 0), weights_seed_crit.fieldname],
+    )
```
Comment (ConnectedSystems) [inline src/io/sampling.jl:56]: Is this the most readable approach?

**Comment by Rosejoycrocker** [comment]
I thinkI've addressed your comments @ConnectedSystems

**Review by ConnectedSystems — CHANGES_REQUESTED** [review]
Sorry, I forgot to hit the submit button for these comments

File: src/decision/dMCDA.jl [hunk 6]
```diff
@@ -182,6 +187,14 @@ function mcda_normalize(x::Matrix)::Matrix
...
 end
 
+"""
+    mcda_normalize(x::DataFrame)::DataFrame
+
+Normalize weights for a set of scenarios (wse/wsh) for MCDA.
+"""
+function mcda_normalize(x::DataFrame)::DataFrame
+    return x ./ sum.(eachrow(x))
+end
```
[the remark this answers is not in the document]
Comment (Rosejoycrocker) [reply src/decision/dMCDA.jl]: Weird, tests are passing for me. WHat you've linked looks different from the current implementation- have you pulled the most recent version?

File: src/interventions/Interventions.jl [hunk 7]
```diff
@@ -117,6 +129,7 @@ Base.@kwdef struct Intervention{N,P,P2} <: EcoModel
         bounds=(2.0, 25.0 + 1.0),
         default_bounds=(2.0, 25.0 + 1.0),
         dists="unif",
+        criteria_keywords=(""),
```
Comment (ConnectedSystems) [inline src/interventions/Interventions.jl:132]: You're aware that this creates an empty string?

If you want a tuple with a single element, it is `("", )`

File: src/io/sampling.jl [hunk 8]
```diff
@@ -41,7 +44,17 @@ function adjust_samples(spec::DataFrame, df::DataFrame)::DataFrame
...
+
+    # Normalize MCDA weights for fogging scenarios
+    is_guided = df.guided .> 0
+    df[(df.fogging .> 0.0) .& is_guided, weights_fog_crit.fieldname] .= mcda_normalize(
+        df[(df.fogging .> 0.0) .& is_guided, weights_fog_crit.fieldname],
+    )
+    # Normalize MCDA weights for seeding scenarios
+    df[.!(not_seeded) .& is_guided, weights_seed_crit.fieldname] .= mcda_normalize(
+        df[.!(not_seeded) .& is_guided, weights_seed_crit.fieldname],
+    )
```
Comment (ConnectedSystems) [inline src/io/sampling.jl:57]: Line 46 still needs to be addressed.

I also don't consider the DataFrame selectors to be very readable but if you're okay with it then I'll live.

File: src/decision/dMCDA.jl [hunk 9]
```diff
@@ -182,6 +187,15 @@ function mcda_normalize(x::Matrix)::Matrix
...
 
+"""
+    mcda_normalize(x::DataFrame)::DataFrame
+
+Normalize weights for a set of scenarios (wse/wsh) for MCDA.
+"""
+function mcda_normalize(x::DataFrame)::DataFrame
+    norm_x = sum.(eachrow(x))
+    return x ./ norm_x
+end
```
Comment (ConnectedSystems) [inline src/decision/dMCDA.jl:198]: Did you compare this with the previous approach converting to matrix?

File: src/io/sampling.jl (hunk 8 above)
Comment (Rosejoycrocker) [reply src/io/sampling.jl]: I've changed them again

File: src/decision/dMCDA.jl (hunk 9 above)
Comment (Rosejoycrocker) [reply src/decision/dMCDA.jl:198]: The Matrix version took 0.406 seconds, the eachrow() version took 0.082 seconds.

As I said in my email, I think this difference is due to the converison to Matrix taking time. I'm guessing eachrow is also optimised for use with dataframes.

Comment (ConnectedSystems) [reply src/decision/dMCDA.jl:198]: How did you do the timing and did you check allocations?

Comment (Rosejoycrocker) [reply src/decision/dMCDA.jl:198]: I assigned both versions to a function and did @time func(x)... I know its preferable to use Benchmarking Tools but I just did it quickly

Comment (ConnectedSystems) [reply src/decision/dMCDA.jl:198]: If you're okay with that then let's roll with it

Comment (Rosejoycrocker) [reply src/decision/dMCDA.jl:198]: I'm okay with it, but are you?

File: src/io/sampling.jl [hunk 10]
```diff
@@ -41,7 +44,18 @@ function adjust_samples(spec::DataFrame, df::DataFrame)::DataFrame
...
-    df[not_fogged, contains.(names(df), "shade_")] .= 0.0
+    df[not_fogged, contains.(names(df), "fog_") .& contains.(names(df), "shade_")] .= 0.0
+
+    # Normalize MCDA weights for fogging scenarios
+    guided_fogged = (df.fogging .> 0.0) .& (df.guided .> 0)
+    df[guided_fogged, weights_fog_crit.fieldname] .= mcda_normalize(
+        df[guided_fogged, weights_fog_crit.fieldname],
+    )
+    # Normalize MCDA weights for seeding scenarios
+    guided_seeded = .!(not_seeded) .& is_guided
```
Comment (ConnectedSystems) [inline src/io/sampling.jl:55]: `is_guided` does not exist any more.

Tests fail due to this.

Comment (Rosejoycrocker) [reply src/io/sampling.jl]: Fixed now, sorry, and tests are passing on my end

**Review by ConnectedSystems — APPROVED** [review]
Tests pass after rebase and scenario runs and figure creation work without issue.

**Pull request merged**